[Addon Manager] Gui Redesign and add support for package.xml metadata

chennes opened the pull request:
This package of changes significantly refactors the Addon Manager to support the [new package.xml metadata format](https://wiki.freecadweb.org/Package_Metadata), as well as adding support in core FreeCAD for workbenches installed into package subdirectories. Most of the changes are behind-the-scenes, but user-visible changes are the switch from tabs to a drop-down box, the addition of a filter box, and a redesigned layout for the packages list.

Please back up your USER_DATA_DIR before testing! 

What needs testing:
* It's pure Python, so there's no compilation in the addon manager, but there are a few changes to the internal metadata code, which is C++, so make sure that compiles on your system
* There are a lot of threads going on in the addon manager, make sure you can open and close it repeatedly without crashes
* Does the revised UI make sense? Does it work well for you? Test out some regular expressions in the filter.
* Test installing a workbench
* Test uninstalling a workbench
* Test installing a macro
* Test uninstalling a macro
* Right now one WB author has implemented the new metadata file: if you want to see the UI when metadata is present, check out the Curves Workbench (thanks @tomate44 !). That's still a work in progress, please let me know your feedback.

You are of course welcome to give feedback on the code, remembering of course that only about 50% of this is actually new code, the rest is the original code that's been moved around. Diff isn't going to be super useful here!

chennes commented:
@tomate44 Do you have GitPython installed? I found some problems with the code when there was no git library. 

I'm still looking into why the CI test is failing.

tomate44 commented:
Yes, I think it is this package : 
```
python-gitpython                [Installed] 3.1.24-1   community 
    A python library used to interact with Git repositories
```

tomate44 commented:
The freezes I was getting, as explained above, are fixed with the latest changes.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_5131/pipeline.svg"/></a>for feature branch [PR_5131](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131). Pipeline [#399136855 ](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/399136855) was triggered at [9f4c3d7](https://github.com/FreeCAD/FreeCAD/pull/5131/commits/9f4c3d73a1fa919b13a0ea85c30e0f4a677426f0). All CI branch [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_5131/pipeline.svg"/></a> for feature branch [PR_5131](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131). Pipeline [402244390](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/402244390) was triggered at [2b54aff](https://github.com/FreeCAD/FreeCAD/pull/5131/commits/2b54aff9b1ec559807cf1d42a5804c5c4f1badad). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

balrobs commented:
Compilation on Ubuntu 20.04 works fine for me but immediately after launching FreeCAD it crashes with this messages:

`
Program received signal SIGSEGV, Segmentation fault.
#0  /lib/x86_64-linux-gnu/libc.so.6(+0x46210) [0x7fcc1298d210]
#1  /lib/x86_64-linux-gnu/libpyside2.cpython-38-x86_64-linux-gnu.so.5.14(+0x23ce2) [0x7fcc11a86ce2]
#2  /lib/x86_64-linux-gnu/libpyside2.cpython-38-x86_64-linux-gnu.so.5.14(+0x23fbf) [0x7fcc11a86fbf]
#3  /lib/x86_64-linux-gnu/libpyside2.cpython-38-x86_64-linux-gnu.so.5.14(+0x25835) [0x7fcc11a88835]
#4  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x249199) [0x7fcc14150199]
#5  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyObject_MakeTpCall+0xab) [0x7fcc141afafb]
#6  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74df3) [0x7fcc13f7bdf3]
#7  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x7d86) [0x7fcc13f83ef6]
#8  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalCodeWithName+0x8fb) [0x7fcc140d1ecb]
#9  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyEval_EvalCodeEx+0x42) [0x7fcc140d2252]
#10  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x1ce54f) [0x7fcc140d554f]
#11  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x25f43f) [0x7fcc1416643f]
#12  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#13  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x12fd) [0x7fcc13f7d46d]
#14  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalCodeWithName+0x8fb) [0x7fcc140d1ecb]
#15  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyEval_EvalCodeEx+0x42) [0x7fcc140d2252]
#16  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyEval_EvalCode+0x1f) [0x7fcc140d263f]
#17  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x1cfc81) [0x7fcc140d6c81]
#18  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x25f527) [0x7fcc14166527]
#19  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyVectorcall_Call+0x60) [0x7fcc141af810]
#20  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x8dd8) [0x7fcc13f84f48]
#21  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalCodeWithName+0x8fb) [0x7fcc140d1ecb]
#22  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyFunction_Vectorcall+0x94) [0x7fcc141af0f4]
#23  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#24  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x7d86) [0x7fcc13f83ef6]
#25  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x8006b) [0x7fcc13f8706b]
#26  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#27  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0xea8) [0x7fcc13f7d018]
#28  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x8006b) [0x7fcc13f8706b]
#29  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#30  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x12fd) [0x7fcc13f7d46d]
#31  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x8006b) [0x7fcc13f8706b]
#32  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#33  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x12fd) [0x7fcc13f7d46d]
#34  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x8006b) [0x7fcc13f8706b]
#35  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x2a90c8) [0x7fcc141b00c8]
#36  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyObject_CallMethodIdObjArgs+0xf0) [0x7fcc141b0320]
#37  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyImport_ImportModuleLevelObject+0x25e) [0x7fcc13f8fc1e]
#38  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x77d1) [0x7fcc13f83941]
#39  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalCodeWithName+0x8fb) [0x7fcc140d1ecb]
#40  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyEval_EvalCodeEx+0x42) [0x7fcc140d2252]
#41  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyEval_EvalCode+0x1f) [0x7fcc140d263f]
#42  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x1cfc81) [0x7fcc140d6c81]
#43  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x25f527) [0x7fcc14166527]
#44  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyVectorcall_Call+0x60) [0x7fcc141af810]
#45  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x8dd8) [0x7fcc13f84f48]
#46  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalCodeWithName+0x8fb) [0x7fcc140d1ecb]
#47  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyFunction_Vectorcall+0x94) [0x7fcc141af0f4]
#48  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#49  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x7d86) [0x7fcc13f83ef6]
#50  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x8006b) [0x7fcc13f8706b]
#51  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#52  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0xea8) [0x7fcc13f7d018]
#53  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x8006b) [0x7fcc13f8706b]
#54  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#55  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x12fd) [0x7fcc13f7d46d]
#56  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x8006b) [0x7fcc13f8706b]
#57  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#58  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x12fd) [0x7fcc13f7d46d]
#59  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x8006b) [0x7fcc13f8706b]
#60  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x2a90c8) [0x7fcc141b00c8]
#61  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyObject_CallMethodIdObjArgs+0xf0) [0x7fcc141b0320]
#62  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyImport_ImportModuleLevelObject+0x25e) [0x7fcc13f8fc1e]
#63  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x77d1) [0x7fcc13f83941]
#64  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalCodeWithName+0x8fb) [0x7fcc140d1ecb]
#65  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyEval_EvalCodeEx+0x42) [0x7fcc140d2252]
#66  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyEval_EvalCode+0x1f) [0x7fcc140d263f]
#67  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x18c0dc) [0x7fcc140930dc]
#68  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyRun_StringFlags+0x79) [0x7fcc14093429]
#69  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x1cfb60) [0x7fcc140d6b60]
#70  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x25f527) [0x7fcc14166527]
#71  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#72  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x12fd) [0x7fcc13f7d46d]
#73  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalCodeWithName+0x8fb) [0x7fcc140d1ecb]
#74  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyFunction_Vectorcall+0x94) [0x7fcc141af0f4]
#75  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#76  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x12fd) [0x7fcc13f7d46d]
#77  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalCodeWithName+0x8fb) [0x7fcc140d1ecb]
#78  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyFunction_Vectorcall+0x94) [0x7fcc141af0f4]
#79  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x74d6d) [0x7fcc13f7bd6d]
#80  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x12fd) [0x7fcc13f7d46d]
#81  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalCodeWithName+0x8fb) [0x7fcc140d1ecb]
#82  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyEval_EvalCodeEx+0x42) [0x7fcc140d2252]
#83  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyEval_EvalCode+0x1f) [0x7fcc140d263f]
#84  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x18c0dc) [0x7fcc140930dc]
#85  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyRun_StringFlags+0x79) [0x7fcc14093429]
#86  0x7fcc146614d1 in Base::InterpreterSingleton::runString[abi:cxx11](char const*) from /home/balrobs/Desktop/freecad-build/lib/libFreeCADBase.so+0xeb
#87  0x7fcc16888ec8 in Gui::Application::runInitGuiScript() from /home/balrobs/Desktop/freecad-build/lib/libFreeCADGui.so+0x4a
#88  0x7fcc1688aba4 in Gui::Application::runApplication() from /home/blarobs/Desktop/freecad-build/lib/libFreeCADGui.so+0x1cb4
#89  ./FreeCAD(+0xa83c) [0x55e563b0483c]
#90  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7fcc1296e0b3]
#91  ./FreeCAD(+0x9aae) [0x55e563b03aae]
`

Any idea?

chennes commented:
No, that backtrace doesn't give a lot of clues: is this a clean compile, or a build on top of an old one?

balrobs commented:
Thanks for your answer!
It was a clean compile without error messages. I did it twice following wiki indications for Ubuntu with the same results. The same procedure used to work for me in the past and recently for compiling edi271's Next31TechDrawTools branch.
Will investigate further when I have some more time.

chennes commented:
Can you run FreeCAD with the `--write-log` option and post the last few lines of the resulting log file?

balrobs commented:
The last few lines of the log file are:

```
...
Err: #84  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x18c0dc) [0x7f70a5fc50dc]
Err: #85  /lib/x86_64-linux-gnu/libpython3.8.so.1.0(PyRun_StringFlags+0x79) [0x7f70a5fc5429]
Err: #86  0x7f70a65934d1 in Base::InterpreterSingleton::runString[abi:cxx11](char const*) from /home/balrobs/Desktop/freecad-build/lib/libFreeCADBase.so+0xeb
Err: #87  0x7f70a87d2468 in Gui::Application::runInitGuiScript() from /home/balrobs/Desktop/freecad-build/lib/libFreeCADGui.so+0x4a
Err: #88  0x7f70a87d4144 in Gui::Application::runApplication() from /home/balrobs/Desktop/freecad-build/lib/libFreeCADGui.so+0x1cb4
Err: #89  ./FreeCAD(+0xa83c) [0x55fa4689b83c]
Err: #90  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f70a48a00b3]
Err: #91  ./FreeCAD(+0x9aae) [0x55fa4689aaae]
```

ipatch commented:
i'll try and test this out here in a few. for us mac users who are building self contained `FreeCAD.app` bundles. there seems to be some hardcoded paths when building for macos using homebrew. specifically related to the paths of the openssl certs. most people on macos have homebrew setup and installed in `/usr/local` but there are some edge cases where people who have setup homebrew within their respected `$HOME` dir.

if macos homebrew is setup in a non default path, should this PR have issues?

i haven't looked all the files changed or commits, but the two currents issues that i'm running into can be further described in the below two freecad forum posts.

- https://forum.freecadweb.org/viewtopic.php?f=42&t=63471
- https://forum.freecadweb.org/viewtopic.php?f=3&t=42825&start=10#p366114

chennes commented:
> The last few lines of the log file are:

Hah! Sorry, I am stupid. I need the last few lines _before_ the backtrace starts. There are only a few places it could be failing, I'm trying to see the last thing it did successfully.

chennes commented:
> if macos homebrew is setup in a non default path, should this PR have issues?

No paths are hardcoded into the Addon Manager: as long as FreeCAD is able to find its own USER_DATA_DIR, everything _should_ work correctly. Which is not to say that it doesn't warrant testing!

balrobs commented:
> 
> 
> > The last few lines of the log file are:
> 
> Hah! Sorry, I am stupid. I need the last few lines _before_ the backtrace starts. There are only a few places it could be failing, I'm trying to see the last thing it did successfully.

Don't worry! Here are hopefully the lines that can help:

```
...
Log: Init: Running FreeCADGuiInit.py start script...
Log: Init:   Searching modules...
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Show(InitGui.py not found)... ignore
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/TechDraw... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Start... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Test... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/PartDesign... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Plot(InitGui.py not found)... ignore
Err: Program received signal SIGSEGV, Segmentation fault.
...
```
Maybe the problem with InitGui.py ??

chennes commented:
@balrobs I've made a small tweak to the output during startup: can you retest, and paste in the entire log output starting at the line "Init: Searching modules..."?

balrobs commented:
> @balrobs I've made a small tweak to the output during startup: can you retest, and paste in the entire log output starting at the line "Init: Searching modules..."? I don't need the backtrace, just up to the SEGFAULT.

Thanks @chennes for your efforts!
I've repeated the test but the lines starting at "Init: Searching modules..." seem to me the same as those I have already posted. So I start some lines bevor at "Init:   Searching for modules...". Hope it helps.
```
Log: Init:   Searching for modules...
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Show(Init.py not found)... ignore
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/TechDraw... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Start... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Test... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/PartDesign... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Plot(Init.py not found)... ignore
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/AddonManager... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Import... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/MeshPart... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Part... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Raytracing... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Spreadsheet... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Image... done
Log: Drawing became obsolete in 0.17; consider using TechDraw instead.
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Drawing... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Fem... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Robot... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/OpenSCAD... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Arch... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Inspection... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Surface... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Idf... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Points... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Tux(Init.py not found)... ignore
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Path... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Measure... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/ReverseEngineering... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Mesh... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Draft... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Sketcher... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Web... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Material... done
Log: Init:      Initializing /home/balrobs/.FreeCAD/Mod/BIM(Init.py not found)... ignore
Log: Init:      Initializing /home/balrobs/.FreeCAD/Mod/Curves(Init.py not found)... ignore
Log: Init: Initializing freecad.Curves
Log: Init: No init module found in freecad.Curves, skipping
Log: Using /home/balrobs/Desktop/freecad-build/Mod as module path!
Log: System path after init:
Log:    /home/balrobs/Desktop/freecad-build/bin
Log:    /home/balrobs/Desktop/freecad-build/Mod/Show
Log:    /home/balrobs/Desktop/freecad-build/Mod/TechDraw
Log:    /home/balrobs/Desktop/freecad-build/Mod/Start
Log:    /home/balrobs/Desktop/freecad-build/Mod/Test
Log:    /home/balrobs/Desktop/freecad-build/Mod/PartDesign
Log:    /home/balrobs/Desktop/freecad-build/Mod/Plot
Log:    /home/balrobs/Desktop/freecad-build/Mod/AddonManager
Log:    /home/balrobs/Desktop/freecad-build/Mod/Import
Log:    /home/balrobs/Desktop/freecad-build/Mod/MeshPart
Log:    /home/balrobs/Desktop/freecad-build/Mod/Part
Log:    /home/balrobs/Desktop/freecad-build/Mod/Raytracing
Log:    /home/balrobs/Desktop/freecad-build/Mod/Spreadsheet
Log:    /home/balrobs/Desktop/freecad-build/Mod/Image
Log:    /home/balrobs/Desktop/freecad-build/Mod/Drawing
Log:    /home/balrobs/Desktop/freecad-build/Mod/Fem
Log:    /home/balrobs/Desktop/freecad-build/Mod/Robot
Log:    /home/balrobs/Desktop/freecad-build/Mod/OpenSCAD
Log:    /home/balrobs/Desktop/freecad-build/Mod/Arch
Log:    /home/balrobs/Desktop/freecad-build/Mod/Inspection
Log:    /home/balrobs/Desktop/freecad-build/Mod/Surface
Log:    /home/balrobs/Desktop/freecad-build/Mod/Idf
Log:    /home/balrobs/Desktop/freecad-build/Mod/Points
Log:    /home/balrobs/Desktop/freecad-build/Mod/Tux
Log:    /home/balrobs/Desktop/freecad-build/Mod/Path
Log:    /home/balrobs/Desktop/freecad-build/Mod/Measure
Log:    /home/balrobs/Desktop/freecad-build/Mod/ReverseEngineering
Log:    /home/balrobs/Desktop/freecad-build/Mod/Mesh
Log:    /home/balrobs/Desktop/freecad-build/Mod/Draft
Log:    /home/balrobs/Desktop/freecad-build/Mod/Sketcher
Log:    /home/balrobs/Desktop/freecad-build/Mod/Web
Log:    /home/balrobs/Desktop/freecad-build/Mod/Material
Log:    /home/balrobs/.FreeCAD/Mod/BIM
Log:    /home/balrobs/.FreeCAD/Mod/Curves
Log:    /usr/local/sbin
Log:    /usr/local/bin
Log:    /usr/sbin
Log:    /usr/bin
Log:    /sbin
Log:    /bin
Log:    /usr/games
Log:    /usr/local/games
Log:    /snap/bin
Log: Init: App::FreeCADInit.py done
Log: Init: Creating Gui::Application and QApplication
Log: Local server 'FreeCAD' started
Log: OpenGL version is: 3.1 (3.1 Mesa 20.2.6)
Log: Run Gui init script
Log: Init: Running FreeCADGuiInit.py start script...
Log: Init:   Searching modules...
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Show(InitGui.py not found)... ignore
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/TechDraw... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Start... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Test... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/PartDesign... done
Log: Init:      Initializing /home/balrobs/Desktop/freecad-build/Mod/Plot(InitGui.py not found)... ignore
Err: Program received signal SIGSEGV, Segmentation fault.
```

chennes commented:
@balrobs That's very helpful, thanks! It helps me narrow down exactly which code is failing. I'll dig into it.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_5131/pipeline.svg"/></a> for feature branch [PR_5131](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131). Pipeline [408627457](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/408627457) was triggered at [c373289](https://github.com/FreeCAD/FreeCAD/pull/5131/commits/c37328959b2a6333fa5216a91d76fa8f6db53c43). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

chennes commented:
@balrobs I think I have found and fixed the issues on Ubuntu 20.04 -- when you have time, can you please re-test? Thanks!

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_5131/pipeline.svg"/></a> for feature branch [PR_5131](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131). Pipeline [408635919](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/408635919) was triggered at [ba4abc7](https://github.com/FreeCAD/FreeCAD/pull/5131/commits/ba4abc749eb29503a42abdaeabd8fd922e2a6955). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_5131/pipeline.svg"/></a> for feature branch [PR_5131](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131). Pipeline [409255280](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/409255280) was triggered at [e41d0a5](https://github.com/FreeCAD/FreeCAD/pull/5131/commits/e41d0a5ac2542c956b54a3d7be520e5b2578c4f5). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

balrobs commented:
> @balrobs I think I have found and fixed the issues on Ubuntu 20.04 -- when you have time, can you please re-test? Thanks!

@chennes amazing, now Freecad no longer crashes on startup :-)
However lunching the addon manager I noticed another problem. It took a very long time (2min 42sec) until the progressbar shows 100% and the workbench list were no longer grayed out.
This is my report view output, where the found metadata and the necessary loading times can be found:
```
23:27:23  flag
23:27:54  Found a metadata file for ThreadProfile
23:27:54  Found a metadata file for Curves
23:27:54  Found a metadata file for DynamicData
23:29:16  Found a metadata file for MeshRemodel
```

chennes commented:
Yes, I agree that's too long, even for a one-time cache update. I have a very fast internet connection, which hides the issue, so thank you for your tests. I am going to refactor so that everything but the initial list-fetch happens in the background.

balrobs commented:
Indeed my slow internet connection at home is a pitty (ca. 10MBit/s).
Thank you for your efforts to improve the user experience for users with poor internet connection.

chennes commented:
@balrobs I have refactored the whole thing and implemented much more caching (with some user preferences for update frequency) -- do you have time to give it another spin?

@tomate44 I have redesigned the GUI -- do you have time to look at it again?

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_5131/pipeline.svg"/></a> for feature branch [PR_5131](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131). Pipeline [423701029](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/423701029) was triggered at [1563c97](https://github.com/FreeCAD/FreeCAD/pull/5131/commits/1563c977a778edc2742faeba9608a4e01e7a9d5b). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

balrobs commented:
@chennes Thanks for your amazing work!
Now all loading stuff is done in the background while the FreeCAD remains responsive ;-)
Unfortunately when I try to install a workbench or a macro I get this yellow warning and nothing else happens:
`01:02:53  Internal error: no selected repo`

tomate44 commented:
Sorry, I don't have much free time right now.
But I compiled and tested yesterday and got same error as balrobs (no selected repo).

chennes commented:
Thank you both again for your testing: I focused so much on the caching that I broke other things, sorry for that. I've gone back and retested the install and uninstall and they should both be working now, so I appreciate the time you took here, it's a great help.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_5131/pipeline.svg"/></a> for feature branch [PR_5131](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131). Pipeline [425424873](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/425424873) was triggered at [5c0fd94](https://github.com/FreeCAD/FreeCAD/pull/5131/commits/5c0fd942e1f3b21402c71522992943e0e9dcf47e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

balrobs commented:
Hi @chennes,

I did some further testing and and experienced this:
- when I click on the close-button while the Addon Manager is loading, the dialog closes and FreeCAD freezes. The mouse pointer remains moveable but FreeCAD doesn't respond to actions any more.
- when I start the Addon Manager, the dialog is painted in black for a very short time before showing his content. This isn't a big deal but quite annoying.
- when I install a workbench, no progressbar is shown any more. As a result users have no feedback if pressing the install-button triggered an action.
- when I try to install the Reinforcement-Workbench, I see a lot of errors in the 
report-view. I ecountered this behaviuor only for this specific workbench. Is this normal?
```
17:17:25  Error loading https://camo.githubusercontent.com/44032dbbaa9d943ff884925474661e51d46e3d5e2b9ebed1f67d94c5b5bba3ef/68747470733a2f2f736f757263657265722e696f2f66616d652f537572616a44616472616c2f616d726974333730312f467265654341442d5265696e666f7263656d656e742f696d616765732f30:
 Not Found
17:17:25  Error loading https://camo.githubusercontent.com/5751a7b21e5278ae60b5127afb32fbbb7d9c226dc282ba77590acae2cc51dafe/68747470733a2f2f736f757263657265722e696f2f66616d652f537572616a44616472616c2f616d726974333730312f467265654341442d5265696e666f7263656d656e742f696d616765732f31:
 Not Found
17:17:25  Error loading https://camo.githubusercontent.com/18d587dbb4a5f7de19cda5001b3c0ed54a6b9024dbd4013d148a77ad5c6d383e/68747470733a2f2f736f757263657265722e696f2f66616d652f537572616a44616472616c2f616d726974333730312f467265654341442d5265696e666f7263656d656e742f696d616765732f32:
 Not Found
17:17:25  Error loading https://camo.githubusercontent.com/5b850d37d4a78c02c0a59489fbc351cebfa091483471d7edf950c678bfd0ad26/68747470733a2f2f736f757263657265722e696f2f66616d652f537572616a44616472616c2f616d726974333730312f467265654341442d5265696e666f7263656d656e742f696d616765732f33:
 Not Found
17:17:26  Error loading https://camo.githubusercontent.com/ed89899429dc00480a0048fe68454b15c6d02e403111c81e5e5cc9bee0fdf44f/68747470733a2f2f736f757263657265722e696f2f66616d652f537572616a44616472616c2f616d726974333730312f467265654341442d5265696e666f7263656d656e742f696d616765732f34:
 Not Found
17:17:26  Error loading https://camo.githubusercontent.com/f0838e18954f1624a6555ee9a288e718ffb828e91d209e40b8cdf30aa3ac14d9/68747470733a2f2f736f757263657265722e696f2f66616d652f537572616a44616472616c2f616d726974333730312f467265654341442d5265696e666f7263656d656e742f696d616765732f35:
 Not Found
17:17:26  Error loading https://camo.githubusercontent.com/96e9337f26cdc4f8bf3da8819dce4521146e86ecb1aea10736294232c3bea7d9/68747470733a2f2f736f757263657265722e696f2f66616d652f537572616a44616472616c2f616d726974333730312f467265654341442d5265696e666f7263656d656e742f696d616765732f36:
 Not Found
17:17:26  Error loading https://camo.githubusercontent.com/c90447dcedfa78c5a0ec9c6599e2d5f6bc4dc958f413af804c49667b07e544bf/68747470733a2f2f736f757263657265722e696f2f66616d652f537572616a44616472616c2f616d726974333730312f467265654341442d5265696e666f7263656d656e742f696d616765732f37:
 Not Found
```

Thanks again for all your work on the Addon Manager!

chennes commented:
> * when I click on the close-button while the Addon Manager is loading, the dialog closes and FreeCAD freezes. The mouse pointer remains moveable but FreeCAD doesn't respond to actions any more.
This proved tricky to resolve, but I think I've got it. Can you re-test?

> * when I start the Addon Manager, the dialog is painted in black for a very short time before showing his content. This isn't a big deal but quite annoying.
I will continue to work on startup speed, but probably won't hold the merge on this point.

> * when I install a workbench, no progressbar is shown any more. As a result users have no feedback if pressing the install-button triggered an action.
I have to rework that code so there is something meaningful to show.

> * when I try to install the Reinforcement-Workbench, I see a lot of errors in the
>   report-view. I ecountered this behaviuor only for this specific workbench. Is this normal?
That seems to be a specific problem with that workbench, I will discuss it with the author.

balrobs commented:
> > * when I click on the close-button while the Addon Manager is loading, the dialog closes and FreeCAD freezes. The mouse pointer remains moveable but FreeCAD doesn't respond to actions any more.
> 
> This proved tricky to resolve, but I think I've got it. Can you re-test?

Yeah ... you solved it ;-)
Now here the problem is gone.

 > > * when I install a workbench, no progressbar is shown any more. As a result users have no feedback if pressing the install-button triggered an action.
>
> I have to rework that code so there is something meaningful to show.

IMHO this would be nice to have since the old Addon Manager had this feature too.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_5131/pipeline.svg"/></a> for feature branch [PR_5131](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131). Pipeline [429647356](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/429647356) was triggered at [8de1a8f](https://github.com/FreeCAD/FreeCAD/pull/5131/commits/8de1a8fee7d242e57e42ae0c39e3af905ea00778). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_5131/pipeline.svg"/></a> for feature branch [PR_5131](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5131). Pipeline [432253337](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/432253337) was triggered at [ec047fd](https://github.com/FreeCAD/FreeCAD/pull/5131/commits/ec047fdd0e59ee1ce616d46ddf4c611c23a7cde4). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).
